### PR TITLE
1033 - Add list answer type for Activity node type

### PR DESCRIPTION
**Pull request by odotadot**
## Changes
- Adds list option for Text questions
- Allow users to set min/max # of fields for list answer type
- Answer node displays list answers correctly

## Screenshot

https://user-images.githubusercontent.com/34692466/124803173-a2925500-df0d-11eb-97c6-69a3fd74459d.mp4

image
image

## Issue Linkage
Closes (issue #1033)
## PR Dependency
Depends on: (PR #1043)
## Automated Testing
- Cannot add new fields if max # of fields already
- Cannot remove fields if min # of fields already
- Can display list answer in Answer node

Tasks - 
- [x] Should only save answers if user press submit
- [x] Merge and then Re-base PR to answer-type

### CODE REVIEW

File: templates/vue/src/components/Lightbox/media/ActivityMedia/Question.vue [hunk 1]
```diff
@@ -1,15 +1,11 @@
...
-      class="button-nav m-auto"
-      @click="back"
-    >
+    <button v-if="formOpened" class="button-nav m-auto" @click="back">
       <i class="fas fa-arrow-left"></i>
     </button>
     <loading v-if="submitting" label="Submitting..." />
     <div v-else>
-      <div v-if="question.followUp.enabled" class="follow-up">
+      <div v-if="question.followUp.questionId !== null" class="follow-up">
```
Comment (EldadZZipori) [inline templates/vue/src/components/Lightbox/media/ActivityMedia/Question.vue:8]: First line should be the correct check

File: templates/vue/src/components/Lightbox/media/ActivityMedia/Question.vue [hunk 2]
```diff
@@ -1,15 +1,11 @@
 <template>
   <div class="question">
-    <button
-      v-if="formOpened && enabledAnswerTypes.length > 1"
-      class="button-nav m-auto"
-      @click="back"
-    >
+    <button v-if="formOpened" class="button-nav m-auto" @click="back">
```
Comment (EldadZZipori) [inline templates/vue/src/components/Lightbox/media/ActivityMedia/Question.vue:3]: First line should be the correct check

File: templates/vue/src/components/Lightbox/media/ActivityMedia/Question.vue [hunk 3]
```diff
@@ -217,33 +230,57 @@ export default {
...
+    const enabledAnswerTypes = Object.entries(this.question.answerTypes)
+      .filter(([, value]) => {
+        return value.enabled
+      })
+      .map(item => item[0])
+
+    if (enabledAnswerTypes.length === 1) {
+      this.formType = enabledAnswerTypes[0]
+      this.formOpened = true
+    }
```
Comment (EldadZZipori) [inline templates/vue/src/components/Lightbox/media/ActivityMedia/Question.vue:276]: Use of openFormIfSingle is correct

File: templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue [hunk 4]
```diff
@@ -39,25 +39,18 @@
...
           >
             <b-form-group>
-              <b-form-checkbox
-                v-model="question.followUp.enabled"
-                switch
-                @change="togglingFollowUp(question)"
-              >
-                {{ question.followUp.enabled ? "Yes" : "No" }}
+              <b-form-checkbox v-model="question.isFollowUp" switch>
+                {{ question.isFollowUp ? "Yes" : "No" }}
```
Comment (EldadZZipori) [inline templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue:43]: Use of question.followUp.enabled is correct

File: templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue [hunk 5]
```diff
@@ -39,25 +39,18 @@
...
               </b-form-checkbox>
             </b-form-group>
-            <b-form-group
-              v-if="question.followUp.enabled"
-              label="Show this text first:"
-            >
+            <b-form-group v-if="question.isFollowUp" label="Show this text first:">
```
Comment (EldadZZipori) [inline templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue:46]: Use of question.followUp.enabled is correct

File: templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue [hunk 6]
```diff
@@ -39,25 +39,18 @@
...
               <b-form-input
                 v-model="question.followUp.text"
                 placeholder="Previously, you said:"
               ></b-form-input>
             </b-form-group>
             <b-form-group
-              v-if="question.followUp.enabled"
+              v-if="question.isFollowUp"
```
Comment (EldadZZipori) [inline templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue:53]: Use of question.followUp.enabled is correct

File: templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue [hunk 7]
```diff
@@ -180,7 +190,6 @@ import RichTextForm from "./RichTextForm"
 const defaultQuestion = {
   text: "",
   followUp: {
-    enabled: false,
```
Comment (EldadZZipori) [inline templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue:183]: we need this

File: templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue [hunk 8]
```diff
@@ -252,11 +265,6 @@ export default {
     getGroupTitle(question, index) {
       return `Question #${index + 1}: ${question.text || "Untitled"}`
     },
-    togglingFollowUp(question) {
-      question.followUp.text = ""
-      question.followUp.questionId = ""
-      question.followUp.nodeId = ""
-    },
```
Comment (EldadZZipori) [inline templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue:259]: restore this function

File: templates/vue/src/store/mutations.js [hunk 9]
```diff
@@ -131,10 +131,6 @@ export function completeQuestion(state, { nodeId, questionId, answerType, answer
     state.userAnswers[nodeId].activity[questionId] = { answers: {} }
   }
 
-  if (typeof state.userAnswers[nodeId].activity[questionId].answers === "string") {
-    state.userAnswers[nodeId].activity[questionId].answers = {}
-  }
-
```
Comment (EldadZZipori) [inline templates/vue/src/store/mutations.js:137]: restore this as well

File: templates/vue/src/components/Lightbox/media/ActivityMedia/Question.vue (hunk 2 above)
Comment (odotadot) [reply templates/vue/src/components/Lightbox/media/ActivityMedia/Question.vue:3]: Patched

File: templates/vue/src/components/Lightbox/media/ActivityMedia/Question.vue (hunk 1 above)
Comment (odotadot) [reply templates/vue/src/components/Lightbox/media/ActivityMedia/Question.vue:8]: Patched

File: templates/vue/src/components/Lightbox/media/ActivityMedia/Question.vue (hunk 3 above)
Comment (odotadot) [reply templates/vue/src/components/Lightbox/media/ActivityMedia/Question.vue:276]: patched

File: templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue (hunk 4 above)
Comment (odotadot) [reply templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue:43]: patched

File: templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue (hunk 5 above)
Comment (odotadot) [reply templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue:46]: patched

File: templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue (hunk 6 above)
Comment (odotadot) [reply templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue:53]: patched

File: templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue (hunk 7 above)
Comment (odotadot) [reply templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue:183]: patched

File: templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue (hunk 8 above)
Comment (odotadot) [reply templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue:259]: patched

File: templates/vue/src/store/mutations.js (hunk 9 above)
Comment (odotadot) [reply templates/vue/src/store/mutations.js:137]: fixed

**Comment by EldadZZipori** [comment]
- [ ] Should only save answers if user press submit

**Comment by Frozemint** [comment]
The element looks kind of good, and I've made some comments in slack. In the interest of record keeping I'll mirror the comments I've made in slack here.

rn the authoring interface is kind of bugged... this is probably an HTML bug. I also thought the "min"  "max" "placeholder" labels are kind of vague as it is, and I think it's a good idea to change them to something more descriptive like "Minimum items in list", "Maximum items in list", "Placeholder list answer", etc.

The current styling of the plus and minus buttons (in the add to list interface) also could use some work... the disabled button style only darkens the blue and red colour very slightly, making it hard to discern that the button is acutally disabled. Perhaps we could use a more visible style, like not drawing the button altogether.

Otherwise it looks quite good, I'm still looking through the code.

File: templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue [hunk 10]
```diff
@@ -140,6 +140,55 @@
...
+                <div class="list-options">
+                  <div class="checkbox-input">
+                    <div class="list-checkbox">
+                      <label for="min-field">Min</label>
+                    </div>
+                    <div class="list-input">
+                      <b-form-input
+                        id="min-field"
+                        v-model="question.answerTypes.list.minFields"
+                        placeholder="Min # fields"
```
Comment (Frozemint) [inline templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue:161]: If the form input type is number, how come the placeholder is a string?

File: templates/vue/src/components/modals/NodeModal/index.vue [hunk 11]
```diff
@@ -921,18 +905,30 @@ export default {
...
+          let listQuestion = this.node.typeData.activity.questions[0].answerTypes
+            .list
+          if (listQuestion.minFields < 1) {
+            errMsgs.push("Minimum # of fields for list activity must be >= 1")
+          }
+          if (listQuestion.maxFields.enabled) {
+            if (listQuestion.maxFields.value < listQuestion.minFields) {
+              errMsgs.push(
+                "Maximum # of fields for list activity must be >= minumum # fields"
+              )
```
Comment (Frozemint) [inline templates/vue/src/components/modals/NodeModal/index.vue:918]: Similarly, there should be a check for if minvalue > maxvalue

File: templates/vue/src/components/Lightbox/media/ActivityMedia/ListQuestion.vue [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+              ? question.answerTypes.list.placeholder
+              : 'Enter text and press Enter'
+          "
+        ></b-form-input>
+        <b-button
+          :disabled="numOfFields >= maxFields"
+          variant="primary"
+          :data-qa="`list-add-${index}`"
+          :class="{
+            'btn-primary': !(numOfFields >= maxFields),
```
[the remark this answers is not in the document]
Comment (odotadot) [reply templates/vue/src/components/Lightbox/media/ActivityMedia/ListQuestion.vue:23]: good catch

File: templates/vue/src/components/Lightbox/media/ActivityMedia/ListQuestion.vue [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+      required: true,
+    },
+    answers: {
+      type: Array,
+      required: true,
+    },
+  },
+  data() {
+    return {
+      answerList: this.answers.map(x => x),
```
[the remark this answers is not in the document]
Comment (odotadot) [reply templates/vue/src/components/Lightbox/media/ActivityMedia/ListQuestion.vue:73]: I think I had it as `answerList: this.answers` before but that makes answerList a reference to the answers prop, which I didn't want because it messes up the submission (answers will be saved even if user doesn't click submit). The identity map makes a new array that resolves this.

File: templates/vue/src/components/Lightbox/media/ActivityMedia/ListQuestion.vue [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+    }
+  },
+  computed: {
+    minFields() {
+      return parseInt(this.question.answerTypes.list.minFields, 10)
+    },
+    maxFields() {
+      return this.question.answerTypes.list.maxFields.enabled
+        ? parseInt(this.question.answerTypes.list.maxFields.value, 10)
+        : 100000
```
[the remark this answers is not in the document]
Comment (odotadot) [reply templates/vue/src/components/Lightbox/media/ActivityMedia/ListQuestion.vue:83]: Yeah the 100000 was arbitrary. I'll make a const variable change it to 100 and set a default value in the authoring side

File: templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue (hunk 10 above)
Comment (odotadot) [reply templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue:161]: The placeholder is just text to prompt the user to know what to put into the input field when it's empty

File: templates/vue/src/components/modals/NodeModal/index.vue (hunk 11 above)
Comment (odotadot) [reply templates/vue/src/components/modals/NodeModal/index.vue:918]: That check covers the logic of `minvalue > maxvalue`

File: templates/vue/src/components/Lightbox/media/ActivityMedia/Question.vue [hunk 15]
```diff
@@ -53,6 +53,12 @@
             :node="node"
             @submit="handleSubmit"
           />
+          <list-question
+            v-else-if="formType === 'list'"
+            :question="question"
+            :answers="answer.length > 0 ? answer : []"
```
[the remark this answers is not in the document]
Comment (odotadot) [reply templates/vue/src/components/Lightbox/media/ActivityMedia/Question.vue:59]: The check is needed since answer is sometimes undefined. I simplified it further

**Comment by Frozemint** [comment]
This looks good with all the changes, out of abundance of caution I would also add another automated test to add 100 items to the list (the maximum) and that everything still works correctly (since we claimed, by design, the functionality supports a maximum of 100 items). We might also need to quickly test this manually to see that the HTML/CSS design doesnt break or looks excessively bad when this many items are added. Otherwise, I think this is pretty good.

File: templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue [hunk 16]
```diff
@@ -140,6 +140,63 @@
...
+                      </div>
+                    </div>
+                  </div>
+                  <label for="placeholder">
+                    Placeholder for answer fields (optional):
+                  </label>
+                  <b-form-input
+                    id="placeholder"
+                    v-model="question.answerTypes.list.placeholder"
+                    data-qa="question-answer-list-placeholder"
```
Comment (wynnset) [inline templates/vue/src/components/modals/NodeModal/forms/ContentForm/ActivityForm.vue:195]: We should have a placeholder prop on this input with it being whatever the default placeholder value is.

File: templates/vue/src/components/modals/NodeModal/forms/ContentForm/TextForm.vue [hunk 17]
```diff
@@ -19,6 +19,11 @@ export default {
       type: Object,
       required: true,
     },
+    // textInput: {
+    //   type: String,
+    //   required: false,
+    //   default: "",
+    // },
```
Comment (wynnset) [inline templates/vue/src/components/modals/NodeModal/forms/ContentForm/TextForm.vue:26]: Please do not leave commented out code or if you have to, please leave an explanation of why it's commented out.

File: templates/vue/src/components/modals/NodeModal/index.vue [hunk 18]
```diff
@@ -1182,16 +1164,13 @@ table {
...
-
 .indented-options {
   border-left: solid 2px #ccc;
   padding-left: 1em;
 }
-
 button:disabled {
   cursor: not-allowed;
 }
-
```
Comment (wynnset) [inline templates/vue/src/components/modals/NodeModal/index.vue:1194]: Why are we removing all these empty lines suddenly from this file?

File: templates/vue/src/components/modals/NodeModal/index.vue [hunk 19]
```diff
@@ -930,18 +914,35 @@ export default {
         if (!validPreviousAnswers) {
           errMsgs.push("Please select a previous activity to display")
         }
+        if (this.node.typeData.activity.questions[0].answerTypes.list.enabled) {
+          let listQuestion = this.node.typeData.activity.questions[0].answerTypes
+            .list
+          let minValue = parseInt(listQuestion.minFields, 10)
+          if (minValue < 1) {
+            errMsgs.push("Minimum # of fields for list activity must be >= 1")
```
Comment (wynnset) [inline templates/vue/src/components/modals/NodeModal/index.vue:922]: It would be more clear to type the word "number" in fully rather than using the "#" shorthand (here and below / elsewhere)

File: templates/vue/src/store/mutations.js [hunk 20]
```diff
@@ -130,7 +130,6 @@ export function completeQuestion(state, { nodeId, questionId, answerType, answer
   if (state.userAnswers[nodeId].activity[questionId] === undefined) {
     state.userAnswers[nodeId].activity[questionId] = { answers: {} }
   }
-
```
Comment (wynnset) [inline templates/vue/src/store/mutations.js:133]: Again, no reason to remove empty line - this is especially more relevant if you have not touched the file otherwise. Touching files will always increase the chances of merge conflicts, so better to avoid touching them if not necessary and relevant to the PR.

**Review by wynnset — CHANGES_REQUESTED** [review]
I am avoiding leaving CSS feedback as I will do those myself
- [x] I am thinking rather than having a separate list answer type, we can just extend the text answer type and add a toggle "Allow entering multiple values", which would then allow the author to define the min/max if they want to. I am saying this because I am realizing these two answer types are extremely similar and it would make more sense for the author, I believe. It would also allow authors to change answer type from text to list if they want to. On the code side, it would allow us to use the text authoring form if we want to, although we may still want to keep a separate component for the user question (whatever you think is best) but of course it would use the same data.
- [ ] Pressing enter while adding list items currently does not do anything. It should add the item, create a new empty list item (if less than max), and focus on that input

File: templates/vue/src/components/Lightbox/media/ActivityMedia/Question/ListQuestion.vue [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+  },
+  created() {
+    if (this.numOfFields === 0) {
+      for (let i = 0; i < this.minFields; i++) {
+        this.addAnswer()
+      }
+    } else {
+      for (let i = this.numOfFields; i < this.minFields; i++) {
+        this.addAnswer()
+      }
```
[the remark this answers is not in the document]
Comment (odotadot) [reply templates/vue/src/components/Lightbox/media/ActivityMedia/Question/ListQuestion.vue:91]: yup

File: templates/vue/src/components/modals/NodeModal/index.vue (hunk 18 above)
Comment (odotadot) [reply templates/vue/src/components/modals/NodeModal/index.vue]: My bad...I forget how that happened. I reverted the changes.

**Comment by odotadot** [comment]
> It would also allow authors to change answer type from text to list if they want to. On the code side, it would allow us to use the text authoring form if we want to, although we may still want to keep a separate component for the user question (whatever you think is best) but of course it would use the same data.

@wynnset When they switch between list and single/multi-line mode in the authoring side, do we want to retain the previously saved answers or discard them? I'm also thinking it might be easier to merge the ListQuestion and TextQuestion components together, since they now fall under one answerType (text).

**Comment by wynnset** [comment]
> > It would also allow authors to change answer type from text to list if they want to. On the code side, it would allow us to use the text authoring form if we want to, although we may still want to keep a separate component for the user question (whatever you think is best) but of course it would use the same data.
> 
> When they switch between list and single/multi-line mode in the authoring side, do we want to retain the previously saved answers or discard them? I'm also thinking it might be easier to merge the ListQuestion and TextQuestion components together, since they now fall under one answerType (text).

Preferably we would want to keep those answers @odotadot I'm ok if you want to merge those components, as well.

**Comment by odotadot** [comment]
The only tests that are failing are the ones related to the list-answers. It might be more efficient to fix these tests after the review has been finalized.

**Comment by EldadZZipori** [comment]
Need to add data/field validation to node and form:
- I am able to submit an activity with a list question without writing any text to all the items I added. When opening the activity again they are still there and empty. When checking with answer node it seem they are being saved as empty strings.
- I can submit the Activity Form with an empty "Maximum answer fields" and "Minimum answer fields" values. They remain empty when re-editing the node/ unchecking the "Maximum answer fields".

I also believe the cypress activity test needs to be edited for the new changes in this PR as it always fails because cypress cannot find the id "question-answer-list"

Other then that everything seems to work. Limits work great. Showing previous answers/ answer node.

**Comment by EldadZZipori** [comment]
@odotadot 100 items in the cypress test is really a lot I think. Would it be of to reduce it to 3-10, that way we also check that the user can actually click the text field there as well.

File: templates/vue/src/components/modals/NodeModal/index.vue [hunk 22]
```diff
@@ -965,7 +965,33 @@ export default {
...
+            if (!Number.isInteger(maxValue)) {
+              errMsgs.push("Maximum number of fields must be an integer value")
+            }
+            if (maxValue < minValue) {
+              errMsgs.push(
+                "Maximum number of fields for list activity must be >= minimum number fields"
+              )
+            }
+            if (maxValue > 100) {
+              errMsgs.push("Maximum number of fields must be <= 100")
```
Comment (EldadZZipori) [inline templates/vue/src/components/modals/NodeModal/index.vue:991]: I think we should make these messages more user friendly, without <= signs or the words "fields" and "integer.
"Minimum number of fields must be an integer value" can be "Please enter a number as the minimum number of items"

**Review by wynnset — APPROVED** [review]

**Comment by wynnset** [comment]
Hi @odotadot could you please fix the tests again and then mark this as ready for merge and assign it to me please? Thank you!

**Comment by odotadot** [comment]
@wynnset It might be good to take one last scan through the recent changes. The single-line text question disappeared so I re-added it to make the cypress tests that relied on it pass. I also refactored the answer node cypress test as it was outdated.

**Review by wynnset — APPROVED** [review]

**Comment by wynnset** [comment]
@odotadot I'm not sure what was changed since I last looked at this, but the list is no longer allowing me to add more than 1 answer as a user:
https://user-images.githubusercontent.com/8981258/126220777-5402cd43-55ec-4eb8-9ad7-faf6b4a3e6aa.png
image

**Comment by wynnset** [comment]
@odotadot sorry my machine was acting up. I restarted node and it worked perfectly
